sync/mutex/mutex.go

```diff
@@ -45,3 +56,62 @@ func With2[T, A, B any](m *Mutex[T], f func(*T) (A, B)) (A, B) {
 	})
 	return a, b
 }
+
+// With3 is like With1, but the callback returns three values instead of one.
+func With3[T, A, B, C any](m *Mutex[T], f func(*T) (A, B, C)) (A, B, C) {
+	var (
+		a A
+		b B
+		c C
+	)
+	m.With(func(t *T) {
+		a, b, c = f(t)
+	})
+	return a, b, c
+}
+
+// With4 is like With1, but the callback returns four values instead of one.
+func With4[T, A, B, C, D any](m *Mutex[T], f func(*T) (A, B, C, D)) (A, B, C, D) {
+	var (
+		a A
+		b B
+		c C
+		d D
+	)
+	m.With(func(t *T) {
+		a, b, c, d = f(t)
+	})
+	return a, b, c, d
+}
+
+// A Locked[T] is a reference to a value of type T which is guarded by a Mutex,
+// which the caller has acquired.
+type Locked[T any] struct {
+	mu       *Mutex[T]
+	unlocked bool
+}
+
+// Value returns a reference to the protected value. It must not be used after
+// calling Unlock.
+//
+// Value will panic if called after Unlock()
+//
+// We recommend against assigning the result of Value() to a variable; while
+// repeatedly having to write l.Value() is mildly annoying, it makes it
+// much harder to accidentally use the value after unlocking.
+func (l *Locked[T]) Value() *T {
+	if l.unlocked {
+		panic("Called Locked.Value after Unlock.")
+	}
+	return &l.mu.val
+}
+
+// Unlock releases the mutex. Any references to the value obtained via Value
+// must not be used after this is called.
+func (l *Locked[T]) Unlock() {
+	if l.unlocked {
```

Comment:
I think this could be simplified by nilling out the pointer to `*Mutex[T]` after unlocking.

Reply:
Done.